extensions/jms/src/main/java/org/mule/extensions/jms/api/source/JmsMessageListener.java

```diff
@@ -0,0 +1,148 @@
+   * @param contentType    Default contentType if the consumed message doesn't provide one
+   */
+  JmsMessageListener(JmsSession session, JmsConfig config, JmsListenerLock jmsLock, JmsSessionManager sessionManager,
+                     SourceCallback<Object, JmsAttributes> sourceCallback, JmsSupport jmsSupport, AckMode ackMode,
+                     String encoding, String contentType) {
+    this.session = session;
+    this.sourceCallback = sourceCallback;
+    this.jmsLock = jmsLock;
+    this.ackMode = ackMode;
+    this.encoding = encoding;
+    this.contentType = contentType;
+    this.config = config;
+    this.sessionManager = sessionManager;
+    this.jmsSupport = jmsSupport;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void onMessage(Message message) {
+    SourceCallbackContext context = sourceCallback.createContext();
+    evaluateAckAction(message);
+    saveReplyToDestination(message, context);
+    context.addVariable(JMS_LOCK_VAR, jmsLock);
+    context.addVariable(JMS_SESSION_VAR, session);
+    produceMessageResult(message, context, resolveEncoding(message), resolveContentType(message));
+    waitForMessageToBeProcessed(jmsLock);
+  }
+
+  private void evaluateAckAction(Message message) {
+    try {
+      evaluateMessageAck(ackMode, session, message, sessionManager, jmsLock);
+    } catch (JMSException e) {
+      LOGGER.error("An error occurred while processing an incoming message: ", e);
+      sourceCallback.onSourceException(e);
+    }
+  }
+
+  private String resolveContentType(Message message) {
+    // If no explicit content type was provided to the operation, fallback to the
+    // one communicated in the message properties. Finally if no property was set,
+    // use the default one provided by the config
+    return resolveOverride(resolveMessageContentType(message, config.getContentType()), contentType);
+  }
+
+  private String resolveEncoding(Message message) {
+    // If no explicit content type was provided to the operation, fallback to the
+    // one communicated in the message properties. Finally if no property was set,
+    // use the default one provided by the config
+    return resolveOverride(resolveMessageEncoding(message, config.getEncoding()), encoding);
+  }
+
+  private void saveReplyToDestination(Message message, SourceCallbackContext context) {
+    try {
+      Destination replyTo = message.getJMSReplyTo();
+      if (replyTo != null) {
+        context.addVariable(REPLY_TO_DESTINATION_VAR, replyTo);
+      }
+    } catch (JMSException e) {
+      LOGGER.error("An error occurred while obtaining the ReplyTo destination: ", e);
+      sourceCallback
+          .onSourceException(new JmsExtensionException(e, "An error occurred while obtaining the ReplyTo destination: "));
+    }
+  }
+
+  private void produceMessageResult(Message message, SourceCallbackContext context, String encoding, String contentType) {
```

Comment:
non descriptive name. Use something like dispatchMEssage or anything which actually gives the impression that a mule message is being pushed to the pipeline